# Requerimientos para módulo Transmisor (TX)
### Instituto Tecnologico de Costa Rica
### Escuela de Ingeniería Electrónica
### Curso: Taller de Comunicaciones Eléctricas
### Estudiantes:
- Esteban Arias Rojas
- David Herrera Castro
- David Monge Naranjo
- Federico Rivera Moya

In [1]:
import numpy as np
import pandas as pd
from utils import encoder
from time import time

## Lectura de datos

In [2]:
file = "filtro_savitzky.xlsx"

# Leer el archivo Excel
datos_pd = pd.read_excel(file)

# Mostrar los datos leídos
print(datos_pd)

          Datos
0     90.787115
1     91.575105
2     92.242893
3     92.990973
4     93.843062
..          ...
122  126.816691
123  125.965607
124  125.276151
125  124.510648
126  123.919015

[127 rows x 1 columns]


In [3]:
datos_array = np.array(datos_pd)

# Convierte los datos en punto flotante a binario de forma vectorizada
datos_binarios = np.vectorize(encoder.float2bin)(datos_array)

# Concatena todos los datos en un solo str
datos_full = ""
for i in range(len(datos_binarios)):
    datos_full += datos_binarios[i][0]

# Convierte los datos a una lista
datos_list = []
for bit in datos_full:
    datos_list.append(int(bit))

print(datos_list)

[0, 1, 0, 1, 1, 0, 1, 1, 0, 1, 0, 1, 1, 1, 0, 0, 0, 1, 0, 1, 1, 1, 0, 0, 0, 1, 0, 1, 1, 1, 0, 1, 0, 1, 0, 1, 1, 1, 1, 0, 0, 1, 0, 1, 1, 1, 1, 1, 0, 1, 0, 1, 1, 1, 1, 1, 0, 1, 1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 1, 0, 1, 1, 0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 0, 1, 1, 0, 1, 1, 0, 0, 1, 0, 0, 0, 1, 1, 0, 0, 1, 0, 1, 0, 1, 1, 0, 0, 1, 1, 0, 0, 1, 1, 0, 0, 1, 1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 0, 1, 1, 0, 1, 0, 0, 0, 0, 1, 1, 0, 1, 0, 0, 1, 0, 1, 1, 0, 1, 0, 1, 0, 0, 1, 1, 0, 1, 0, 1, 1, 0, 1, 1, 0, 1, 0, 1, 1, 0, 1, 1, 0, 1, 1, 0, 0, 0, 1, 1, 0, 1, 1, 0, 0, 0, 1, 1, 0, 1, 1, 0, 1, 0, 1, 1, 0, 1, 1, 1, 0, 0, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 0, 1, 0, 1, 1, 1, 0, 0, 0, 1, 0, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 1, 0, 1, 1, 1, 0, 1, 0, 0, 0, 1, 1, 1, 0, 1, 0, 1, 0, 1, 1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 0, 1, 1, 1, 0, 1, 1, 1, 1, 0, 0, 0, 0, 1, 1, 1, 1, 0, 0, 0, 0, 1, 1, 1, 1, 0, 0, 1, 0, 1, 1, 1, 1, 

## Variante código Hamming (7, 4)

In [4]:
# Variante (7, 4)

# Longitud del bloque
n1 = 7

# Número de bits de mensaje
k1 = 4

# Matriz de paridad para variante (7, 4)
P1 = np.array([
    [1, 1, 0],
    [0, 1, 1],
    [1, 1, 1],
    [1, 0, 1]
])

I_k1 = np.eye(k1)
I_nk1 = np.eye(n1 - k1)

In [5]:
# Matriz generadora G en el formato [I_k | P] para el código de Hamming (7, 4)
G1 = np.concatenate((I_k1, P1), axis=1).astype(int)

# Matriz de revisión de paridad H
H1 = np.concatenate((np.transpose(P1), I_nk1), axis=1).astype(int)

### Ejemplo

In [6]:
# Datos de entrada (4 bits)
data_bits = [0, 0, 0, 1]

In [7]:
# Codificar los datos
encoded_bits = encoder.hamming_encode(data_bits, G1)

In [8]:
print("Datos originales:", data_bits)
print("Datos codificados:", encoded_bits)

Datos originales: [0, 0, 0, 1]
Datos codificados: [0, 0, 0, 1, 1, 0, 1]


### Datos leídos

In [9]:
datos_codificados = encoder.hamming_encode_block(datos_list, 74, G1)

print(datos_codificados)

[0, 1, 0, 1, 1, 1, 0, 1, 0, 1, 1, 1, 0, 0, 0, 1, 0, 1, 1, 1, 0, 1, 1, 0, 0, 1, 0, 1, 0, 1, 0, 1, 1, 1, 0, 1, 1, 0, 0, 1, 0, 1, 0, 1, 0, 1, 1, 1, 0, 1, 1, 0, 1, 0, 0, 0, 0, 1, 0, 1, 1, 1, 0, 1, 1, 1, 0, 0, 1, 0, 0, 1, 0, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 0, 1, 0, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 1, 1, 0, 1, 0, 1, 1, 0, 1, 0, 0, 0, 0, 1, 0, 1, 1, 1, 0, 1, 1, 0, 1, 0, 0, 0, 0, 1, 1, 0, 1, 0, 0, 1, 1, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 1, 0, 1, 1, 0, 1, 0, 0, 0, 1, 0, 1, 1, 1, 0, 0, 1, 1, 0, 1, 0, 0, 0, 1, 1, 0, 1, 0, 0, 0, 1, 1, 0, 1, 0, 0, 0, 1, 1, 1, 0, 0, 1, 0, 1, 1, 0, 1, 0, 0, 0, 1, 1, 1, 0, 0, 1, 0, 1, 1, 0, 1, 0, 0, 1, 0, 0, 0, 1, 1, 0, 0, 1, 1, 0, 1, 0, 0, 1, 0, 0, 1, 0, 1, 1, 0, 1, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 0, 1, 0, 1, 1, 0, 1, 0, 0, 1, 0, 1, 1, 1, 0, 0, 0, 1, 1, 0, 1, 0, 0, 1, 0, 1, 1, 1, 0, 0, 0, 1, 1, 0, 1, 0, 0, 1, 1, 0, 0, 1, 0, 1, 0, 1, 1, 0, 1, 0, 0, 1, 1, 0, 0, 1, 0, 1, 0, 1, 1, 0, 1, 0, 0, 1, 1, 0, 1, 

## Se guardan datos

In [10]:
type(datos_codificados)

list

In [11]:
frame_74 = pd.DataFrame({"Datos": datos_codificados})

frame_74.to_excel("datos_codificados_74.xlsx", index=False)

## Variante código Hamming (15, 11)

In [12]:
# Variante (15, 11)

# Longitud del bloque
n2 = 15

# Número de bits de mensaje
k2 = 11

# Matriz de paridad para variante (7, 4)
P2 = np.array([
    [1, 1, 1, 1],
    [1, 1, 1, 0],
    [1, 1, 0, 1],
    [1, 1, 0, 0],
    [1, 0, 1, 1],
    [1, 0, 1, 0],
    [1, 0, 0, 1],
    [0, 1, 1, 1],
    [0, 1, 1, 0],
    [0, 1, 0, 1],
    [0, 0, 1, 1],
])

I_k2 = np.eye(k2)
I_nk2 = np.eye(n2 - k2)

In [13]:
# Matriz generadora G en el formato [I_k | P] para el código de Hamming (15, 11)
G2 = np.concatenate((I_k2, P2), axis=1).astype(int)

# Matriz de revisión de paridad H
H2 = np.concatenate((np.transpose(P2), I_nk2), axis=1).astype(int)

### Ejemplo

In [14]:
# Datos de entrada (11 bits)
data_bits = [0, 0, 0, 1, 1, 1, 0, 1, 0, 1, 0]

In [15]:
# Codificar los datos
encoded_bits = encoder.hamming_encode(data_bits, G2)

In [16]:
print("Datos originales:", data_bits)
print("Datos codificados:", encoded_bits)

Datos originales: [0, 0, 0, 1, 1, 1, 0, 1, 0, 1, 0]
Datos codificados: [0, 0, 0, 1, 1, 1, 0, 1, 0, 1, 0, 1, 1, 1, 1]


## Datos leídos

In [17]:
datos_codificados = encoder.hamming_encode_block(datos_list, 1511, G2)

print(datos_codificados)

[0, 1, 0, 1, 1, 0, 1, 1, 0, 1, 0, 0, 0, 1, 0, 1, 1, 1, 0, 0, 0, 1, 0, 1, 1, 1, 0, 1, 0, 1, 0, 0, 0, 1, 0, 1, 1, 1, 0, 1, 0, 1, 1, 0, 1, 1, 0, 1, 1, 1, 1, 0, 0, 1, 0, 1, 1, 0, 1, 0, 1, 1, 1, 1, 0, 1, 0, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 1, 1, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 1, 1, 0, 0, 0, 0, 1, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 0, 1, 1, 1, 0, 0, 1, 0, 1, 1, 0, 0, 1, 0, 0, 0, 1, 1, 1, 1, 1, 1, 0, 0, 1, 0, 1, 0, 1, 1, 0, 0, 1, 1, 0, 1, 1, 1, 0, 0, 1, 1, 0, 0, 1, 1, 1, 0, 1, 1, 0, 0, 1, 1, 0, 0, 1, 1, 1, 0, 1, 1, 0, 1, 0, 1, 0, 1, 0, 0, 0, 0, 1, 1, 0, 1, 0, 0, 1, 0, 1, 0, 1, 0, 1, 1, 0, 1, 0, 1, 0, 0, 1, 0, 0, 0, 0, 1, 0, 1, 0, 1, 1, 0, 1, 1, 0, 1, 0, 0, 0, 1, 0, 1, 1, 0, 1, 1, 0, 1, 1, 0, 0, 0, 0, 1, 1, 0, 1, 1, 0, 1, 1, 0, 0, 0, 1, 1, 0, 1, 0, 0, 0, 1, 1, 0, 1, 0, 1, 1, 0, 1, 1, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0, 1, 1, 1, 1, 0, 0, 1, 0, 1, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 0, 1, 0, 1, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 1, 1, 0, 0, 0, 1, 0, 1, 1, 1, 1, 1, 1, 1, 0, 

## Se guardan datos

In [18]:
frame_1511 = pd.DataFrame({"Datos": datos_codificados})

frame_1511.to_excel("datos_codificados_1511.xlsx", index=False)

# Tiempo de ejecución

In [19]:
numero_ejecuciones = 200

## Variante código (7, 4)

In [20]:
inicio = time()

for i in range(numero_ejecuciones):
    datos_codificados = encoder.hamming_encode_block(datos_list, 74, G1)

fin = time()

tiempo_74 = (fin - inicio) / 200

## Variante código (15, 11)

In [21]:
inicio = time()

for i in range(numero_ejecuciones):
    datos_codificados = encoder.hamming_encode_block(datos_list, 1511, G2)

fin = time()

tiempo_1511 = (fin - inicio) / 200

## Resultados

In [22]:
print('Tiempo promedio usando condificación (7, 4): ', f'{tiempo_74*1e6:.4f} \u03BCs')
print('Tiempo promedio usando condificación (15, 11): ', f'{tiempo_1511*1e6:.4f} \u03BCs')

Tiempo promedio usando condificación (7, 4):  1828.3045 μs
Tiempo promedio usando condificación (15, 11):  979.9087 μs


# Análisis de variantes de código

## Tabla comparativa

|                            | Código (7, 4)                                                | Código (15, 11)                                              |
| :-------------------------- | :----------------------------------------------------------- | :------------------------------------------------------------ |
| Razón de código            | 4/7 = 0.57. La cantidad de bits de un mensaje aumenta aproximadamente 75% | 11/15 = 0.73. la cantidad de bits de un mensaje aumenta aproximadamente 36% |
| Complejidad implementación | Se deben codificar solamente 4 bits, tanto en hardware o software se realizan menos operaciones, esto conlleva a menor consumo de recursos. Pero para tramas largas de datos, se deben realizar más codificaciones. | Se codifican mayor cantidad de bits, lo que aumenta el consumo de recursos, pero para tramas largas de datos, hay a la larga menos bloques que codificar. |
| Tiempo ejecución           | 1828.3045 μs                                                 | 979.9087 μs                                                  |
| Corrección de errores      | Se pueden detectar 2 errores y corregir 1. Pero cada bloque tiene 7 bits, con una menor cantidad de bits por bloque, hay menor probabilidad de errores | Se pueden detectar 2 errores y corregir 1. Pero para un bloque de 15 bits, hay mayor cantidad de bits que pueden presentar errores y se mantiene la cantidad de errores que se pueden corregir y detectar |